### PR TITLE
Fix race(deadlock) between waitgroup and returnCh

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
wg.Wait() and returnCh is listening in the cases f same `select`. Consider the following scenario:
1. Two pods are being drained in two different goroutines
2. global timeout has hit, which closes stopCh and wg.Wait() waits in the select case
3. one of the pods gets evicted successfuly and wants to write `returnCh <- nil`.

deadlocked. Wait() keeps waiting for all the goroutines and goroutine keeps waiting to write on returnCh(which is blocked because select is stuck in the case where wg.wait() is waiting and case for listening returnCh never gets executed) 

Issue got caught in the CI test on the PR which was trying to revendor kubernetes-drain at aws actuator repo: https://github.com/openshift/cluster-api-provider-aws/pull/243

### CODE REVIEW

File: drain.go [hunk 1]
```diff
@@ -474,12 +470,8 @@ func evictPods(client typedpolicyv1beta1.PolicyV1beta1Interface, pods []corev1.P
 		case <-globalTimeoutCh:
 			logf(options.Logger, "Closing stopCh")
 			close(stopCh)
-			wg.Wait()
```
Comment (smarterclayton) [inline drain.go:477]: Where is the stopCh actually used?

Comment (vikaschoudhary16) [reply drain.go:477]: here, https://github.com/openshift/kubernetes-drain/pull/4/files#diff-b35cc60e9fab98b13f2a1f43795d4d00R429, to exit goroutine

**Comment by eparis** [comment]
@vikaschoudhary16 can we please get at least SOME CI on this repo? pls?

**Comment by eparis** [comment]
@vikaschoudhary16 can we please get at least SOME CI on this repo? pls?

also, this code is a fork from upstream isn't it? From the upstream kubectl drain code, no? Do these still exist upstream?

**Comment by vikaschoudhary16** [comment]
> @vikaschoudhary16 can we please get at least SOME CI on this repo? pls?
> 
> also, this code is a fork from upstream isn't it? From the upstream kubectl drain code, no? Do these still exist upstream?

Yes, copied at some point from the upstream kubectl drain code. That code still exists upstream. Efforts are going on to copy this code into upstream kubernetes-sigs/cluster-api. 
https://github.com/kubernetes/kubernetes/pull/80045/files
https://github.com/kubernetes-sigs/cluster-api/pull/1096/files

Now i checked, this issue of "goroutines not exiting after global timeout in case of eviction   failure" is there in upstream kubectl drain code as well.

Eventually we will move to using drain code from upstream cluster-api. Do you think we can manage without CI by that time?

**Comment by vikaschoudhary16** [comment]
or may be i think, we can have this drain logic in our vendored cluster-api post this PR(because this is needed for an urgent BZ) and make actuators use drain logic from there. CI tests also run on vendored cluster-api.

**Comment by enxebre** [comment]
This repo has no owner, CI, or release branching enable https://github.com/openshift/kubernetes-drain. I'd suggest we snapshot the code as a lib in our cluster-api

**Comment by vikaschoudhary16** [comment]
Created Issue and PR at upstream:
https://github.com/kubernetes/kubernetes/pull/81336/files
https://github.com/kubernetes/kubernetes/issues/81333

**Comment by enxebre** [comment]
We moved the draining code to live in tree in the `openshift/cluster-api` until we eventually move to consume from upstream where this has been reported https://github.com/kubernetes/kubernetes/issues/81333
https://github.com/kubernetes/kubernetes/pull/81336
This patch already sneaked in in the relevant PR https://github.com/openshift/cluster-api/pull/62.
@smarterclayton can we merge this patch for removing skew and then possibly archive this repo to avoid confusion. I don't have the perms to do any of those.

**Comment by TwiN** [comment]
Any chance this will be merged?

**Pull request closed without merging**